### ISSUE TITLE
Cling (rightfully) confused about forward-declared template specializations

### ISSUE THREAD

**Issue by hahnjo** [root-project/root#13815]
### Description

Cling is (rightfully) confused when evaluating a type trait on a forward-declared template specialization and then supplying the full definition (see reproducer). This breaks `roottest-root-meta-ROOT-7462-make` with the upgrade to LLVM 16 (https://github.com/root-project/root/pull/13273) in a C++20 build with GCC 13. This is because the test first expects `std::pair<std::string, Inner<int>>` to be "bad" with only a forward declaration and then to succeed after providing the definition. This doesn't work with the `requires` implementation of `libstdc++` under C++20:
```
In file included from instlib dictionary payload:5:
./instGood.cxx:10:33: error: no matching constructor for initialization of 'std::pair<string, Inner<int> >' (aka 'pair<basic_string<char>, Inner<int> >')
   std::pair<string,Inner<int>> p;
                                ^
/usr/lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/stl_pair.h:249:7: note: candidate constructor not viable: constraints not satisfied
      pair()
      ^
/usr/lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/stl_pair.h:251:12: note: because 'is_default_constructible_v<Inner<int> >' evaluated to false
               && is_default_constructible_v<_T2>
                  ^
```

(I'm still investigating why this doesn't happen in `master` with LLVM 13...)

### Reproducer

```
root [0] template <typename T> class Inner;
root [1] template <> class Inner<int>;
root [2] std::is_default_constructible_v<Inner<int>>
(... error: incomplete type 'Inner<int>' used in type trait expression ...)
root [3] template <> class Inner<int> { public: int fValue; };
root [4] std::is_default_constructible_v<Inner<int>>
(const bool) false
```

Interestingly enough, the following works:
```
root [0] struct A;
root [1] std::is_default_constructible_v<A>
(... error: incomplete type 'A' used in type trait expression ...)
root [2] struct A { int v; };
root [3] std::is_default_constructible_v<A>
(const bool) true
```

**Comment by hahnjo** [comment]
> I'm still investigating why this doesn't happen in `master` with LLVM 13...

Oh well, this can be summarized as "sheer luck". Let me explain: In `master`, LLVM 13 only checks the `explicit` clause when trying to construct the `std::pair`:
```
root [0] template <typename T> class Inner;
root [1] template <> class Inner<int>;
root [2] std::pair<int, Inner<int>> p1;
In module 'std' imported from input_line_1:1:
/opt/gcc/13.2.0/include/c++/13.2.0/bits/stl_pair.h:194:11: error: field has incomplete type 'Inner<int>'
      _T2 second;                ///< The second member
          ^
ROOT_prompt_2:1:28: note: in instantiation of template class 'std::pair<int, Inner<int> >' requested here
std::pair<int, Inner<int>> p1;
                           ^
ROOT_prompt_1:1:19: note: forward declaration of 'Inner<int>'
template <> class Inner<int>;
                  ^
In module 'std' imported from input_line_1:1:
/opt/gcc/13.2.0/include/c++/13.2.0/type_traits:1040:25: error: incomplete type 'Inner<int>' used in type trait expression
      = __bool_constant<__is_constructible(_Tp, _Args...)>;
                        ^
/opt/gcc/13.2.0/include/c++/13.2.0/type_traits:1265:21: note: in instantiation of template type alias '__is_constructible_impl' requested here
    : public __and_<__is_constructible_impl<_Tp>,
                    ^
/opt/gcc/13.2.0/include/c++/13.2.0/type_traits:161:30: note: in instantiation of template class 'std::__is_implicitly_default_constructible<Inner<int> >' requested here
                                      __enable_if_t<bool(_Bn::value)>...>;
                                                         ^
/opt/gcc/13.2.0/include/c++/13.2.0/type_traits:177:16: note: while substituting explicitly-specified template arguments into function template '__and_fn' 
    : decltype(__detail::__and_fn<_Bn...>(0))
               ^
/opt/gcc/13.2.0/include/c++/13.2.0/type_traits:182:29: note: in instantiation of template class 'std::__and_<std::__is_implicitly_default_constructible<int>, std::__is_implicitly_default_constructible<Inner<int> > >' requested here
    : __bool_constant<!bool(_Pp::value)>
                            ^
/opt/gcc/13.2.0/include/c++/13.2.0/bits/stl_pair.h:247:16: note: in instantiation of template class 'std::__not_<std::__and_<std::__is_implicitly_default_constructible<int>, std::__is_implicitly_default_constructible<Inner<int> > > >' requested here
      explicit(__not_<__and_<__is_implicitly_default_constructible<_T1>,
               ^
ROOT_prompt_2:1:28: note: in instantiation of template class 'std::pair<int, Inner<int> >' requested here
std::pair<int, Inner<int>> p1;
                           ^
ROOT_prompt_1:1:19: note: forward declaration of 'Inner<int>'
template <> class Inner<int>;
                  ^
```
Consequently `std::is_default_constructible_v<Inner<int>>` is not instantiated and the construction succeeds after providing the full definition:
```
root [3] template <> class Inner<int> { public: int fValue; };
root [4] std::pair<int, Inner<int>> p2;
```

With LLVM 16, Clang *also* checks the `requires` constraints immediately:
```
root [0] template <typename T> class Inner;
root [1] template <> class Inner<int>;
root [2] std::pair<int, Inner<int>> p1;
In module 'std' imported from input_line_1:1:
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:194:11: error: field has incomplete type 'Inner<int>'
      _T2 second;                ///< The second member
          ^
ROOT_prompt_2:1:28: note: in instantiation of template class 'std::pair<int, Inner<int> >' requested here
std::pair<int, Inner<int>> p1;
                           ^
ROOT_prompt_1:1:19: note: forward declaration of 'Inner<int>'
template <> class Inner<int>;
                  ^
In module 'std' imported from input_line_1:1:
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/type_traits:1040:25: error: incomplete type 'Inner<int>' used in type trait expression
      = __bool_constant<__is_constructible(_Tp, _Args...)>;
                        ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/type_traits:1265:21: note: in instantiation of template type alias '__is_constructible_impl' requested here
    : public __and_<__is_constructible_impl<_Tp>,
                    ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/type_traits:161:30: note: in instantiation of template class 'std::__is_implicitly_default_constructible<Inner<int> >' requested here                                                                                                                                                                      
                                      __enable_if_t<bool(_Bn::value)>...>;
                                                         ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/type_traits:177:16: note: while substituting explicitly-specified template arguments into function template '__and_fn'                                                                                                                                                                                     
    : decltype(__detail::__and_fn<_Bn...>(0))
               ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/type_traits:182:29: note: in instantiation of template class 'std::__and_<std::__is_implicitly_default_constructible<int>, std::__is_implicitly_default_constructible<Inner<int> > >' requested here                                                                                                       
    : __bool_constant<!bool(_Pp::value)>
                            ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:247:16: note: in instantiation of template class 'std::__not_<std::__and_<std::__is_implicitly_default_constructible<int>, std::__is_implicitly_default_constructible<Inner<int> > > >' requested here                                                                                     
      explicit(__not_<__and_<__is_implicitly_default_constructible<_T1>,
               ^
ROOT_prompt_2:1:28: note: in instantiation of template class 'std::pair<int, Inner<int> >' requested here
std::pair<int, Inner<int>> p1;
                           ^
ROOT_prompt_1:1:19: note: forward declaration of 'Inner<int>'
template <> class Inner<int>;
                  ^
In module 'std' imported from input_line_1:1:
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/type_traits:3251:54: error: incomplete type 'Inner<int>' used in type trait expression
  inline constexpr bool is_default_constructible_v = __is_constructible(_Tp);
                                                     ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:251:12: note: in instantiation of variable template specialization 'std::is_default_constructible_v<Inner<int> >' requested here                                                                                                                                                           
               && is_default_constructible_v<_T2>
                  ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:251:12: note: while substituting template arguments into constraint expression here
               && is_default_constructible_v<_T2>
                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:249:7: note: while checking constraint satisfaction for function 'pair' required here
      pair()
      ^~~~
ROOT_prompt_2:1:28: note: in instantiation of template class 'std::pair<int, Inner<int> >' requested here
std::pair<int, Inner<int>> p1;
                           ^
ROOT_prompt_1:1:19: note: forward declaration of 'Inner<int>'
template <> class Inner<int>;
                  ^
(... more errors ...)
```
Consequently `std::is_default_constructible_v<Inner<int>>` is instantiated and we run into the problem described above, even after providing the full definition:
```
root [3] template <> class Inner<int> { public: int fValue; };
root [4] std::pair<int, Inner<int>> p2;
ROOT_prompt_4:1:28: error: no matching constructor for initialization of 'std::pair<int, Inner<int> >'
std::pair<int, Inner<int>> p2;
                           ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:249:7: note: candidate constructor not viable: constraints not satisfied
      pair()
      ^
/opt/gcc/13.2.0/lib/gcc/x86_64-pc-linux-gnu/13.2.0/../../../../include/c++/13.2.0/bits/stl_pair.h:251:12: note: because 'is_default_constructible_v<Inner<int> >' evaluated to false
               && is_default_constructible_v<_T2>
                  ^
(... more errors ...)
```

To prove the point, it is possible to provoke the problem on `master` as well, by manually instantiating `std::is_default_constructible_v<Inner<int>>`:
```
root [0] template <typename T> class Inner;
root [1] template <> class Inner<int>;
root [2] std::is_default_constructible_v<Inner<int>>
In module 'std' imported from input_line_1:1:
/opt/gcc/13.2.0/include/c++/13.2.0/type_traits:3251:54: error: incomplete type 'Inner<int>' used in type trait expression
  inline constexpr bool is_default_constructible_v = __is_constructible(_Tp);
                                                     ^
ROOT_prompt_2:1:6: note: in instantiation of variable template specialization 'std::is_default_constructible_v<Inner<int> >' requested here
std::is_default_constructible_v<Inner<int>>
     ^
ROOT_prompt_1:1:19: note: forward declaration of 'Inner<int>'
template <> class Inner<int>;
                  ^
root [3] template <> class Inner<int> { public: int fValue; };
root [4] std::pair<int, Inner<int>> p2;
ROOT_prompt_4:1:28: error: no matching constructor for initialization of 'std::pair<int, Inner<int> >'
std::pair<int, Inner<int>> p2;
                           ^
/opt/gcc/13.2.0/include/c++/13.2.0/bits/stl_pair.h:249:7: note: candidate constructor not viable: constraints not satisfied
      pair()
      ^
/opt/gcc/13.2.0/include/c++/13.2.0/bits/stl_pair.h:251:12: note: because 'is_default_constructible_v<Inner<int> >' evaluated to false
               && is_default_constructible_v<_T2>
                  ^
(... more errors ...)
```

**Comment by Axel-Naumann** [comment]
@pcanal FYI we need your verdict here.

**Comment by pcanal** [comment]
See also https://github.com/root-project/root/issues/13461.
This requires serious attention to the forward declaration writer or it requires to make the use of module/pcm (instead of the rootmap) file practical.   See at least https://github.com/root-project/root/issues/13543 and https://github.com/root-project/root/issues/13481

**Comment by Axel-Naumann** [comment]
Sorry, @pcanal - that's way too terse for me. Can you expand what a seemingly broken test design has to do with these issues? (Other than the fact that this is all about dictionaries.)

**Comment by pcanal** [comment]
>  has to do with these issues?

Nothing.  You are right. I completely misread the text when responding.

TLDR: Read just the last paragraph.

>  Can you expand what a seemingly broken test design

Can you tell me what is broken in the test design?

It is testing a "natural" control flow for a REPL:
```
First Load a bunch of stuff that, in this case, include a forward declaration.

Second do:
   if (some entity (directly or indirectly) that happens to have a forward declaration is not defined  yet)
       load the definition for the entity.
```
(here the entity is `std::pair<int, Inner<int>>` and the forward declared type is `Inner<int>`)

The test is properly revealing that (as described by Jonas) this pattern is now broken because some intermediary template does not get recalculated or reinstantiated when the definition is loaded.

One could be tempted to 'improve' the test with a less intrusive way of checking but this would likely render the test useless for its intended purpose.  In the real use case those same test are buried deep in the Core/Meta and I/O and should then be changed there too (assuming it is 'possible' to do so).

For more details on the original problem see: 
   https://sft.its.cern.ch/jira/browse/ROOT-7462
and detailed analyis:
   https://github.com/root-project/roottest/blob/master/root/meta/ROOT-7462/README
   
So most likely than not we need to find a way to solve this within Cling and/or Core/Meta.

One hint is that the reproducer symptoms as described by Jonas seem familiar.  For example, it shows a failure to instantiate the template `std::is_default_constructible_v`.  When this is triggered by Core/Meta/Cling, eg `ClassInfo` the error is (supposed to be) supressed **AND** the failed template instantiation is supposed to be cleaned up and removed.  It seems that the problem might be that with Clang16 some side-effects of the failed instantiation are no longer fully removed.

**Comment by pcanal** [comment]
Fair enough.  The LLVM 16 is revealing a part of the problem we had not solved yet :) [Note the simplified example is different from the real use case in that there is no expectation of recovery in that case while inside Core/Meta there is; i.e. once the problem is fixed, the command line reproducer will (and should) still fail ]

**Comment by hahnjo** [comment]
> Actually with the date we have so far, we do not know whether the problem is pre-exiting or not. To know we would need to use (in the same code pattern as the test) a class that triggers the same time of side effect (template instantiation) as is done by LLVM 16 with the pair.

Yes, valid point. The following should replicate the sequence of events and already fails in ROOT v6.20 with LLVM 5:
```
root [0] namespace NS { template <typename T> struct A; }
root [1] template <typename T> struct B { static_assert(std::is_default_constructible_v<T>); };
root [2] TClass::GetClass("B<NS::A<int>>")
(TClass *) nullptr
root [3] namespace NS { template <> struct A<int> { int v; }; }
root [4] B<NS::A<int>> b;
error: static_assert failed
fatal error: cannot open file 'input_line_114': No such file or directory
ROOT_prompt_4:1:15: note: in instantiation of template class '__cling_N50::B<NS::A<int> >' requested here
B<NS::A<int>> b;
              ^
```
whereas the instantiation works fine without the call to `TClass::GetClass`. (note that the `namespace NS` is needed to avoid bogus `'A' does not name a template but is followed by template arguments` in v6.22 and older)

**Comment by hahnjo** [comment]
> Either way, it seems that the issue is that the transaction unloader is not unloading enough.

I've been trying for a bit, and it's actually not too hard to make the `DeclUnloader` reset a `VarDecl`'s `EvaluatedStmt`:
```diff
diff --git a/interpreter/cling/lib/Interpreter/DeclUnloader.cpp b/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
index 25c01b1ef3..004ee9eaff 100644
--- a/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
+++ b/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
@@ -662,6 +662,14 @@ namespace cling {
       MaybeRemoveDeclFromModule(GD);
     }
 
+    // VD->dump();
+    if (auto *Evaluated = VD->getEvaluatedStmt()) {
+      llvm::errs() << "DeclUnloader: " << VD->getNameAsString() << "\n";
+      Evaluated->WasEvaluated = false;
+      Evaluated->Evaluated = APValue();
+      // VD->dump();
+    }
+
     // VarDecl : DeclaratiorDecl, Redeclarable
     bool Successful = VisitRedeclarable(VD, VD->getDeclContext());
     Successful &= VisitDeclaratorDecl(VD);
```

This ensures that the initialization is re-evaluated the next time. Unfortunately this is not sufficient because the `TypeTraitExpr` has its value passed at construction time, so the result is just the same as before.

**Comment by pcanal** [comment]
@vgvassilev Do you have any idea of what else could be missing from the `DeclUnloader` (and/or the transaction decl capture mechanism) to allow this to work?

**Comment by vgvassilev** [comment]
It seems we fail to react on the fact that a forward declaration requires complete type. Without a debugger at hand, it looks like clang misses some callback to `Sema::RequireCompleteType` at the place where we instantiate the template argument. How did you conclude that the issue is in the DeclUnloader?

**Comment by pcanal** [comment]
In either the original test (via `TClingClassInfo`) or the latest example `TClass::GetClass("B<NS::A<int>>")`, the code triggers attempts to instantiate a template which fails (due to the type being only forward declared).

The later error messages (after the full declaration is loaded) indicates that some of the traits template instance are NOT being re-instantiated based using the complete information (and instead the one based on the incomplete information are used).

Technically, we don't know yet whether it is a problem with the transaction unloader or the the transaction decl capture (or possibly even on the failure detection).

What we do know is that the desired behavior [in last example](https://github.com/root-project/root/issues/13815#issuecomment-1759250811) is that after the execution of  `TClass::GetClass("B<NS::A<int>>")` (i.e at the end of line 2), `Cling` should NOT be holding an instantiation of  the trait (not sure which)

**Comment by hahnjo** [comment]
I agree that there is a problem here. That might involve the `DeclCollector` or the `DeclUnloader` or some other parts of ROOT. However, IMHO https://github.com/root-project/root/issues/13815#issuecomment-1759250811 clearly shows that the underlying problem is much bigger and older than just the failing test with LLVM 16. So I'd be very much interested in a pragmatic solution to resolve this last blocker for the upgrade.

**Comment by vgvassilev** [comment]
I agree this should not hold off the llvm16 upgrade assuming we can move forward by disabling the test and work on a fix at the same time.

**Comment by pcanal** [comment]
> One idea I had was to split the test into two parts, ie one "bad" part that tests the behavior with an incomplete class and one "good" part that loads inst2lib.

I am not following the details but the nature of the test is to see what happens if you late load the right declaration after the partial one has been used/tickled, so It sounds like splitting the test would render it 'moot'.

**Comment by pcanal** [comment]
> IMHO https://github.com/root-project/root/issues/13815#issuecomment-1759250811 clearly shows that the underlying problem

I would agree if it was not the fact that this is being triggered by a benign and wildely used class ... i.e. `std::pair` ... 

Consequently, it my opinion we are almost guaranteed to have this problem appear almost immediately in the wild (i.e. code and code flow that worked before will now fail)

**Comment by vgvassilev** [comment]
There types of things are the reason we went to modules and are advocating people to transition. This infrastructure cannot work at 100% and it is pretty much best effort.

**Comment by pcanal** [comment]
>  we went to modules and are advocating people to transition. 

Too bad there is a few more things to fix up to make the transition easy (see Chris Green experience).

> These types of things are the reason we went to module

Humm ... this would not address this ... this is a case of incremental development (eg. we have not migrated `ACLiC` to use module yet)

**Comment by vgvassilev** [comment]
> > we went to modules and are advocating people to transition.
> 
> Too bad there is a few more things to fix up to make the transition easy (see Chris Green experience).
> 

There are things to do to improve it but none seem major blockers. 

> > These types of things are the reason we went to module
> 
> Humm ... this would not address this ... this is a case of incremental development (eg. we have not migrated `ACLiC` to use module yet)

ACLiC supports modules pretty well if we build with modules on.

**Comment by pcanal** [comment]
> ACLiC supports modules pretty well if we build with modules on

Is that code already uploaded?  How do you turn it on? (I.e. we are talking about ACLiC producing modules, right?)

> There are things to do to improve it but none seem major blockers.

They deterred at least one user from proceeding.

Anyway, modules are not at this moment widely used (CMSSW is NOT using them in production) and thus are not a solution to this problem.   If the PR we are talking is merge as-is, it will be a regression (test will need to be disabled and real user cases will stop working).  I assume it will be fine to merge it nonetheless **as long as** this issue is resolved before is a release a made (i.e. this issue and re-enabling the test will be a blocker for the release).  I would prefer that this is investigated before merging though.

**Comment by vgvassilev** [comment]
> > ACLiC supports modules pretty well if we build with modules on
> 
> Is that code already uploaded? How do you turn it on? (I.e. we are talking about ACLiC producing modules, right?)

This code has been there for years. Yea aclic can produce modules and modulemaps. 

> 
> > There are things to do to improve it but none seem major blockers.
> 
> They deterred at least one user from proceeding.
> 

I was not aware they planned to move to modules and I was not aware they gave up. 

> Anyway, modules are not at this moment widely used (CMSSW is NOT using them in production) and thus are not a solution to this problem. If the PR we are talking is merge as-is, it will be a regression (test will need to be disabled and real user cases will stop working). I assume it will be fine to merge it nonetheless **as long as** this issue is resolved before is a release a made (i.e. this issue and re-enabling the test will be a blocker for the release). I would prefer that this is investigated before merging though.

I do not oppose to what you say here. However, as I said, modules are designed to solve these issues which are next to impossible to fix. If we can fix it that’s great but my general r recommendation is to move to modules and reach out if there are problems.
